drivers/liebert-gxe.c

```diff
@@ -0,0 +1,466 @@
+		dstate_datastale();
+		return;
+	}
+
+	poll_state = GXE_WARNING;
+
+	/* Field 6, Nominal Voltage */
+	dstate_setinfo("output.voltage.nominal", "%d",
+		       ydn23_val_from_hex(YDN23_FRAME_REG(frame, 9), 4));
+	/* Field 7, Nominal Frequency */
+	dstate_setinfo("output.frequency.nominal", "%d",
+		       ydn23_val_from_hex(YDN23_FRAME_REG(frame, 11), 4));
+	/* Field 10, Bypass Working Voltage Max, ALWAYS 115% */
+	if (ydn23_val_from_hex(YDN23_FRAME_REG(frame, 17), 4) == 1)
+		dstate_setinfo("input.transfer.bypass.high", "%f",
+			       ydn23_val_from_hex(YDN23_FRAME_REG(frame, 9), 4)*1.15f);
+	/* Field 11, Bypass Working Voltage Min, Volt */
+	if (ydn23_val_from_hex(YDN23_FRAME_REG(frame, 19), 4) == 1)
+		dstate_setinfo("input.transfer.bypass.low", "%d", 120);
+	/* Field 21, Battery Test Interval, per 3 mons */
+	dstate_setinfo("ups.test.interval", "%lu",
+		       (long) ydn23_val_from_hex(YDN23_FRAME_REG(frame, 39), 4)*3*108000);
+
+	dstate_dataok();
+}
+
+static void upsdrv_updateinfo_warning(void)
+{
+	struct ydn23_frame frame;
+	int ret, val;
+	size_t i;
+
+	ydn23_frame_init(&frame, YDN23_GET_WARNING_DATA,
+			 "21", devaddr, NULL, 0);
+
+	ret = ydn23_frame_send(upsfd, &frame);
+	if (ret <= 0) {
+		upslogx(LOG_WARNING, "failed sending WARNING request, retry");
+		dstate_datastale();
+		return;
+	}
+
+	ret = ydn23_frame_read(upsfd, &frame, 1, 0);
+	if (ret <= 0) {
+		upslogx(LOG_WARNING, "failed reading WARNING data, retry");
+		dstate_datastale();
+		return;
+	}
+
+	poll_state = GXE_ONOFF;
+
+	alarm_init();
+	for (i = 0; i < ARRAY_SIZE(gxe_warns); i++) {
+		if (!gxe_warns[i])
+			continue;
+		val = ydn23_val_from_hex(YDN23_FRAME_REG(frame, 1+i), 2);
+		switch(val)  {
+		case 0x00:
+			break;
+		case 0x01:
+		case 0x02:
+		case 0x03:
+		case 0xf0:
+			alarm_set(gxe_warns[i]);
+			break;
+		default:
+			upslogx(LOG_WARNING, "unexpected warning val %x", val);
+			break;
+		}
+	}
+	alarm_commit();
+
+	dstate_dataok();
+}
+
+void upsdrv_updateinfo(void)
+{
+	switch(poll_state) {
+	case GXE_ANALOG:
+		upslogx(LOG_DEBUG, "Polling ANALOG data");
+		upsdrv_updateinfo_analog();
+		break;
+	case GXE_ONOFF:
+		upslogx(LOG_DEBUG, "Polling ONOFF data");
+		upsdrv_updateinfo_onoff();
+		break;
+	case GXE_WARNING:
+		upslogx(LOG_DEBUG, "Polling WARNING data");
+		upsdrv_updateinfo_warning();
+		break;
+	case GXE_SYSPARAM:
+		upslogx(LOG_DEBUG, "Polling SYSPARAM data");
+		upsdrv_updateinfo_sysparam();
+		break;
+	}
+}
+
+void upsdrv_initinfo(void)
+{
+	struct ydn23_frame sendframe, recvframe;
+	char databuf[11];
+	int retry, ret;
+
+	ydn23_frame_init(&sendframe, YDN23_GET_VENDOR_INFO,
+			 "21", devaddr, NULL, 0);
+
+	for (retry = 0; retry < PROBE_RETRIES; retry++) {
+		ret = ydn23_frame_send(upsfd, &sendframe);
+		if (ret <= 0) continue;
+		ret = ydn23_frame_read(upsfd, &recvframe, 1, 0);
+		if (ret > 0)
+			break;
+	}
+
+	if (ret <= 0)
+		fatal_with_errno(EXIT_FAILURE, "gxe: failed reading response");
```

Comment:
Saw this complaint pattern recently, not sure OTOH in which PR:

````
[1mliebert-gxe.c:411:6: [0m[0;1;31merror: [0m[1mvariable 'ret' may be uninitialized when used here [-Werror,-Wconditional-uninitialized][0m
        if (ret <= 0)
[0;1;32m            ^~~
[0m[1mliebert-gxe.c:398:16: [0m[0;1;30mnote: [0minitialize the variable 'ret' to silence this warning[0m
        int retry, ret;
[0;1;32m                      ^
[0m[0;32m                       = 0
[0m1 error generated.
make[1]: *** [Makefile:2113: liebert-gxe.o] Error 1
````

It seems `clang` static analysis (across a ton of compiler versions, 7 to 18 at least) can not guarantee that we would generally get into the loop and so assign `ret`, although it seems obvious to human eye that we should. So it asks to pre-initialize e.g. `int ret = -1` in line 398 to surely use a deterministic value in line 411.

I'd shrug it off as a compiler bug, but:
* it bites us in CI which we want to pass warnings-free anyway
* it does not hurt much to add the assignment (just annoys to do so :) )
* maybe the compiler does know better about how it optimizes something away...

Reply:
Posted the fix commit.